File: csharp/133_Clone_Graph.cs
// DFS
public class Solution {
    Dictionary<Node, Node> m = new Dictionary<Node, Node>();

    public Node CloneGraph(Node node) {
        if (node == null) return null;

        if (!m.ContainsKey(node)) {
            m[node] = new Node(node.val);
            foreach (Node neighbor in node.neighbors) {
                m[node].neighbors.Add(CloneGraph(neighbor));
            }
        }

        return m[node];
    }
}

// BFS
public class Solution {
    public Node CloneGraph(Node node) {
        if (node == null) return null;

        Dictionary<Node, Node> m = new Dictionary<Node, Node>();
        Queue<Node> q = new Queue<Node>();

        Node root = new Node(node.val);
        m.Add(node, root);
        q.Enqueue(node);

        while (q.Count > 0) {
            Node curr = q.Dequeue();

            foreach (Node neighbor in curr.neighbors) {
                // node not existed
                if (!m.ContainsKey(neighbor)) {
                    Node clone = new Node(neighbor.val);
                    m.Add(neighbor, clone);
                    q.Enqueue(neighbor);
                }

                m[curr].neighbors.Add(m[neighbor]);
            }
        }

        return root;
    }
}

File: cpp/133_Clone_Graph.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    vector<Node*> neighbors;
    Node() {
        val = 0;
        neighbors = vector<Node*>();
    }
    Node(int _val) {
        val = _val;
        neighbors = vector<Node*>();
    }
    Node(int _val, vector<Node*> _neighbors) {
        val = _val;
        neighbors = _neighbors;
    }
};
*/

class Solution {
public:
    Node* cloneGraph(Node* node) {
        if (node == nullptr) return nullptr;

        unordered_map<int, Node*> adj, visited;
        queue<Node*> q;

        // clone first node
        visitNode(adj, node, visited, q);

        // traverse all node
        while (!q.empty()) {
            visitNode(adj, q.front(), visited, q);
            q.pop();
        }

        // return adj.begin()->second;
        return adj[1];
    }

    void visitNode(unordered_map<int, Node*>& adj, Node* node, 
                    unordered_map<int, Node*>& visited, queue<Node*>& q) {        
        createNodeIfNeeded(adj, node->val);
        Node *temp = adj[node->val];

        // add adj node to this node
        for (Node* neighbor : node->neighbors) {
            createNodeIfNeeded(adj, neighbor->val);

            addNeighborToNode(temp, adj[neighbor->val]);

            // add node to queqe if not visited
            if (visited.find(neighbor->val) == visited.end())
                q.push(neighbor);
        }

        visited[temp->val] = temp;
    }

    void createNodeIfNeeded(unordered_map<int, Node*>& adj, int val) {
        if (adj.find(val) == adj.end()) {
            Node *temp = new Node(val);
            adj[val] = temp;
        }
    }

    void addNeighborToNode(Node* node, Node* addedNode) {
        // dont add if existed
        for (Node *temp : node->neighbors) {
            if (temp->val == addedNode-> val) return;
        }
        node->neighbors.push_back(addedNode);
    }
};